File: leetcode/13. Roman to Integer.cpp
class Solution {
public:
    int romanToInt(string s) {
        int last = 0;
        int res = 0;
        for(auto c : s){
            int num;
            switch(c){
                case 'I':
                    num = 1;
                    break;
                case 'V':
                    num = 5;
                    break;
                case 'X':
                    num = 10;
                    break;
                case 'L':
                    num = 50;
                    break;
                case 'C':
                    num = 100;
                    break;
                case 'D':
                    num = 500;
                    break;
                default:
                    num = 1000;
            }
            res += num;
            if(num > last){
                res -= 2 * last;
            }
            last = num;
        }
        return res;
    }
};

File: leetcode/13. Roman to Integer.rs
use std::collections::HashMap;

impl Solution {
    pub fn roman_to_int(s: String) -> i32 {
        let m = HashMap::from([
            ('I', 1),
            ('V', 5),
            ('X', 10),
            ('L', 50),
            ('C', 100),
            ('D', 500),
            ('M', 1000),
        ]);
        
        let mut last = ' ';
        let mut res = 0;
        for c in s.chars() {
            res += m[&c];
            if last != ' ' && m[&c] > m[&last] {
                res -= 2 * m[&last];
            }
            last = c;
        }
        
        return res;
    }
}
